modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java

```diff
@@ -17,94 +17,128 @@
-    /** {@link SnapshotStorage} instance for the partition. */
-    private final PartitionSnapshotStorage snapshotStorage;
+    private static final long NETWORK_TIMEOUT = 10_000;
+
+    private final PartitionSnapshotStorage partitionSnapshotStorage;
 
-    /** Snapshot URI. */
     private final SnapshotUri snapshotUri;
 
-    /** Rebalance thread-pool, used to write data into a storage. */
-    //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-    // Use external pool.
-    private final ExecutorService threadPool = Executors.newSingleThreadExecutor();
+    private final IgniteSpinBusyLock cancelBusyLock = new IgniteSpinBusyLock();
+
+    private final CompletableFuture<?> future = new CompletableFuture<>();
 
     /**
      * Snapshot meta read from the leader.
      *
      * @see SnapshotMetaRequest
      */
-    private SnapshotMeta snapshotMeta;
+    @Nullable
+    private volatile SnapshotMeta snapshotMeta;
 
     /**
      * Constructor.
      *
-     * @param snapshotStorage Snapshot storage.
+     * @param partitionSnapshotStorage Snapshot storage.
      * @param snapshotUri Snapshot URI.
      */
-    public IncomingSnapshotCopier(PartitionSnapshotStorage snapshotStorage, SnapshotUri snapshotUri) {
-        this.snapshotStorage = snapshotStorage;
+    public IncomingSnapshotCopier(PartitionSnapshotStorage partitionSnapshotStorage, SnapshotUri snapshotUri) {
+        this.partitionSnapshotStorage = partitionSnapshotStorage;
         this.snapshotUri = snapshotUri;
     }
 
     @Override
-    public void cancel() {
-        //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-        // Implement.
+    public void start() {
+        runAsync(this::start0, partitionSnapshotStorage.getIncomingSnapshotsExecutor());
     }
 
     @Override
     public void join() throws InterruptedException {
-        //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-        // Implement proper join.
+        try {
+            future.get();
+        } catch (CancellationException e) {
+            LOG.info("Copier is cancelled");
+
+            if (!isOk()) {
+                setError(RaftError.ECANCELED, "Copier has been cancelled");
+            }
+        } catch (ExecutionException e) {
+            Throwable cause = e.getCause();
+
+            LOG.error("Error when completing the copier", cause);
+
+            if (cause instanceof SnapshotCopierException) {
+                SnapshotCopierException snapshotCopierException = (SnapshotCopierException) cause;
+
+                if (!isOk()) {
+                    setError(snapshotCopierException.status.getRaftError(), snapshotCopierException.status.getErrorMsg());
+                }
+
+                cause = snapshotCopierException.getCause();
+            } else {
+                if (!isOk()) {
+                    setError(RaftError.UNKNOWN, "Unknown error on completion the copier");
+                }
+            }
+
+            throw new IllegalStateException(cause);
```

Comment:
We should stop using `IllegalStateException`. What does it mean?

Reply:
I think this is ok, it's kind of like `assert false`. This should never happen if there is no bug, right?

Reply:
Tried to fix it.